### PR TITLE
Migrate query tests to the native test runner

### LINKED ISSUE #15455: [core] Add support for running query tests in mbgl-render-test
`mbgl-render-test` should support running query tests

### PR DESCRIPTION
**Pull request by jmalanen**
Closes #15455 
Closes mapbox/mapbox-gl-native-team#37

### CODE REVIEW

File: render-test/parser.cpp [hunk 1]
```diff
@@ -179,8 +184,67 @@ TestPaths makeTestPaths(mbgl::filesystem::path stylePath) {
     };
 }
 
+void writeJSON(rapidjson::PrettyWriter<rapidjson::StringBuffer>& writer, const mbgl::Value& value) {
+    value.match([&](const mbgl::NullValue&) { writer.Null(); },
```
Comment (pozdnyakov) [inline render-test/parser.cpp]: would `[&writer]` suffice?

File: render-test/parser.cpp [hunk 2]
```diff
@@ -179,8 +184,67 @@ TestPaths makeTestPaths(mbgl::filesystem::path stylePath) {
...
+std::string toJSON(const std::vector<mbgl::Feature>& features, unsigned indent, bool singleLine) {
+    rapidjson::CrtAllocator allocator;
+    rapidjson::StringBuffer buffer;
+    rapidjson::PrettyWriter<rapidjson::StringBuffer> writer(buffer);
+    if (singleLine) {
+        writer.SetFormatOptions(rapidjson::kFormatSingleLineArray);
+    }
+    writer.SetIndent(' ', indent);
+    writer.StartArray();
+    for (size_t i = 0; i < features.size(); i++) {
```
Comment (pozdnyakov) [inline render-test/parser.cpp]: nit: ++i

File: render-test/runner.cpp [hunk 3]
```diff
@@ -46,7 +47,236 @@ const std::vector<std::string>& TestRunner::getPlatformExpectationsPaths() {
...
+
+        // We strip precision twice in a row here to avoid cases where
+        // stripping an already stripped number will modify its value
+        // due to bad floating point precision luck
+        // eg `Math.floor(8.16598 * 100000) / 100000` -> 8.16597
+        const double firstStrip = std::floor(*num * multiplier) / multiplier;
+        return std::floor(firstStrip * multiplier) / multiplier;
+    }
+
+    if (value.is<std::vector<Value>>()) {
```
Comment (pozdnyakov) [inline render-test/runner.cpp]: if (value.getVector())

File: render-test/runner.cpp [hunk 4]
```diff
@@ -46,7 +47,236 @@ const std::vector<std::string>& TestRunner::getPlatformExpectationsPaths() {
...
+        std::vector<Value> stripped;
+        const auto& vec = value.get<std::vector<Value>>();
+        stripped.reserve(vec.size());
+        for (const auto& val : vec) {
+            stripped.emplace_back(stripPrecision(val));
+        }
+        return stripped;
+    } else if (value.is<std::unordered_map<std::string, Value>>()) {
```
Comment (pozdnyakov) [inline render-test/runner.cpp]: `value.getObject()`

File: render-test/runner.cpp [hunk 5]
```diff
@@ -46,7 +47,236 @@ const std::vector<std::string>& TestRunner::getPlatformExpectationsPaths() {
...
+        for (std::size_t i = 0; i < avec.size(); ++i) {
+            if (!deepEqual(avec[i], bvec[i])) {
+                return false;
+            }
+        }
+        return true;
+    }
+
+    if (a.is<std::unordered_map<std::string, Value>>()) {
+        const auto& amap = a.get<std::unordered_map<std::string, Value>>();
```
Comment (pozdnyakov) [inline render-test/runner.cpp]: const auto& amap = *a.getObject();

File: render-test/parser.cpp (hunk 1 above)
Comment (jmalanen) [reply render-test/parser.cpp]: Yes, changed.

File: render-test/parser.cpp (hunk 2 above)
Comment (jmalanen) [reply render-test/parser.cpp]: Done

File: render-test/runner.cpp (hunk 4 above)
Comment (jmalanen) [reply render-test/runner.cpp]: Done

File: render-test/runner.cpp (hunk 5 above)
Comment (jmalanen) [reply render-test/runner.cpp]: Done

File: render-test/runner.cpp (hunk 3 above)
Comment (jmalanen) [reply render-test/runner.cpp]: Done - I assume you meant getArray()

File: render-test/runner.cpp [hunk 6]
```diff
@@ -46,7 +47,236 @@ const std::vector<std::string>& TestRunner::getPlatformExpectationsPaths() {
...
+    const auto& bnum = numericValue<double>(b);
+    if (anum && bnum) {
+        return stripPrecision(*anum) == stripPrecision(*bnum);
+    }
+
+    if (a.which() != b.which()) {
+        return false;
+    }
+
+    if (a.is<std::vector<Value>>()) {
```
Comment (pozdnyakov) [inline render-test/runner.cpp]: a.getArray() ? Could you update also other usages?

Comment (jmalanen) [reply render-test/runner.cpp]: Done - missed couple places

File: render-test/runner.cpp [hunk 7]
```diff
@@ -46,7 +47,236 @@ const std::vector<std::string>& TestRunner::getPlatformExpectationsPaths() {
     return result;
 }
 
-bool TestRunner::checkResults(mbgl::PremultipliedImage&& actualImage, TestMetadata& metadata) {
+// Strip precision for numbers, so that we can compare evaluated results with fixtures.
+// Copied from JS expression harness.
+Value stripPrecision(const Value& value) {
```
Comment (alexshalamov) [inline render-test/runner.cpp]: could we share common code between expression and render runners?

File: render-test/runner.cpp [hunk 8]
```diff
@@ -46,7 +47,236 @@ const std::vector<std::string>& TestRunner::getPlatformExpectationsPaths() {
...
+std::vector<std::string> tokenize(std::string str) {
+    std::vector<std::string> tokens;
+    std::regex re("\n");
+    std::copy(std::regex_token_iterator<std::string::iterator>(str.begin(), str.end(), re, -1),
+              std::regex_token_iterator<std::string::iterator>(),
+              std::back_inserter(tokens));
+    return tokens;
+}
+
+std::string simpleDiff(const Value& result, const Value& expected) {
```
Comment (alexshalamov) [inline render-test/runner.cpp]: could it be also shared?

File: render-test/runner.cpp (hunk 7 above)
Comment (jmalanen) [reply render-test/runner.cpp]: Done: https://github.com/mapbox/mapbox-gl-native/pull/15770

**Pull request merged**